File: src/main/java/io/github/romeoahmed/cursedoath/technique/TechniqueProjectile.java
package io.github.romeoahmed.cursedoath.technique;

import net.minecraft.network.syncher.EntityDataAccessor;
import net.minecraft.network.syncher.EntityDataSerializer;
import net.minecraft.network.syncher.EntityDataSerializers;
import net.minecraft.network.syncher.SynchedEntityData;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.InterpolationHandler;
import net.minecraft.world.entity.SteppedInterpolationHandler;
import net.minecraft.world.entity.projectile.Projectile;
import net.minecraft.world.level.Level;
import net.minecraft.world.phys.Vec3;

public abstract sealed class TechniqueProjectile extends Projectile permits TechniqueOrb, TechniqueWave {
    static final EntityDataSerializer<Vec3> LAUNCH_POSITION_SERIALIZER =
            EntityDataSerializer.forValueType(Vec3.STREAM_CODEC);
    private static final EntityDataAccessor<Integer> TECHNIQUE =
            SynchedEntityData.defineId(TechniqueProjectile.class, EntityDataSerializers.INT);
    private static final EntityDataAccessor<Vec3> LAUNCH_POSITION =
            SynchedEntityData.defineId(TechniqueProjectile.class, LAUNCH_POSITION_SERIALIZER);

    protected TechniqueProjectile(EntityType<? extends TechniqueProjectile> type, Level level) {
        super(type, level);
    }

    public final Technique technique() {
        var value = Technique.fromWire(entityData.get(TECHNIQUE));
        return value == null ? Technique.PURPLE : value;
    }

    public final Vec3 launchPosition() {
        return entityData.get(LAUNCH_POSITION);
    }

    @Override
    protected void defineSynchedData(SynchedEntityData.Builder builder) {
        builder.define(TECHNIQUE, Technique.PURPLE.wireId());
        builder.define(LAUNCH_POSITION, Vec3.ZERO);
    }

    @Override
    protected InterpolationHandler createInterpolationHandler() {
        return SteppedInterpolationHandler.create(this);
    }

    protected final void launch(ServerPlayer player, Technique ability) {
        setOwner(player);
        entityData.set(TECHNIQUE, ability.wireId());
        var eye = player.getEyePosition();
        var muzzle = eye.add(player.getLookAngle().scale(TechniqueTuning.launchDistance(ability)));
        entityData.set(LAUNCH_POSITION, muzzle);
        setPos(muzzle);
        setRot(player.getYRot(), player.getXRot());
    }
}

File: src/main/kotlin/io/github/romeoahmed/cursedoath/technique/TechniqueProjectile.kt
package io.github.romeoahmed.cursedoath.technique

import net.minecraft.network.syncher.EntityDataSerializers
import net.minecraft.network.syncher.SynchedEntityData
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.InterpolationHandler
import net.minecraft.world.entity.SteppedInterpolationHandler
import net.minecraft.world.entity.projectile.Projectile
import net.minecraft.world.level.Level

/** Native entity tracking is the single source of position, technique and lifetime for rendering. */
sealed class TechniqueProjectile(
    type: EntityType<out TechniqueProjectile>,
    level: Level,
) : Projectile(type, level) {
    var technique: Technique
        get() = Technique.fromWire(entityData.get(TECHNIQUE)) ?: Technique.PURPLE
        protected set(value) {
            entityData.set(TECHNIQUE, value.wireId)
        }

    override fun defineSynchedData(builder: SynchedEntityData.Builder) {
        builder.define(TECHNIQUE, Technique.PURPLE.wireId)
    }

    override fun createInterpolationHandler(): InterpolationHandler = SteppedInterpolationHandler.create(this)

    protected fun launch(
        player: ServerPlayer,
        ability: Technique,
    ) {
        setOwner(player)
        technique = ability
        setPos(player.eyePosition)
        setRot(player.yRot, player.xRot)
    }

    private companion object {
        val TECHNIQUE = SynchedEntityData.defineId(TechniqueProjectile::class.java, EntityDataSerializers.INT)
    }
}
